doc/source/advancedexamples.rst

```diff
@@ -579,86 +579,149 @@ and using the `Sequential model API <https://www.tensorflow.org/api_docs/python/
+initial density matrix by passing the corresponding array when executing the
+circuit. If a state vector is passed as an initial state in a density matrix
+circuit, it will be transformed automatically to the equivalent density matrix.
 
-The user can simulate noise using :class:`qibo.base.gates.NoiseChannel`.
-If this or any other channel is used in a ``Circuit``, then the execution
-will automatically switch to density matrices. For example
+Qibo currently provides the following channels that can be used during a
+density matrix simulation: :class:`qibo.base.gates.KrausChannel`,
+:class:`qibo.base.gates.UnitaryChannel` and
+:class:`qibo.base.gates.PauliNoiseChannel`.
+We refer to the documentation of each channel for more details.
+Noise can be simulated using these channels, for example:
 
 .. code-block:: python
 
-    import qibo
-    qibo.set_backend("matmuleinsum")
     from qibo import models, gates
 
-    c = models.Circuit(2) # starts with state |00>
-    c.add(gates.X(1)) # transforms |00> -> |01>
-    c.add(gates.NoiseChannel(0, px=0.3)) # transforms |01> -> (1 - px)|01><01| + px |11><11|
+    c = models.Circuit(2, density_matrix=True) # starts with state |00><00|
+    c.add(gates.X(1))
+    # transforms |00><00| -> |01><01|
+    c.add(gates.PauliNoiseChannel(0, px=0.3))
+    # transforms |01><01| -> (1 - px)|01><01| + px |11><11|
     final_state = c()
     # will return tf.Tensor(diag([0, 0.7, 0, 0.3]))
 
 will perform the transformation
 
 .. math::
-    |00\rangle & \rightarrow (I \otimes X)|00\rangle = |01\rangle
+    |00\rangle \langle 00|& \rightarrow (I \otimes X)|00\rangle \langle 00|(I \otimes X)
+    = |01\rangle \langle 01|
     \\& \rightarrow 0.7|01\rangle \langle 01| + 0.3(X\otimes I)|01\rangle \langle 01|(X\otimes I)^\dagger
     \\& = 0.7|01\rangle \langle 01| + 0.3|11\rangle \langle 11|
 
-Note that :class:`qibo.base.circuit.BaseCircuit` will use state vectors until
-the first channel is found and will switch to density matrices for the rest of
-the simulation. Measurements and callbacks can be used exactly as in the case
-of state vector simulation.
+Measurements and callbacks can be used with density matrices exactly as in the
+case of state vector simulation.
+
+
+.. _repeatedexec-example:
+
+Using repeated execution
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Simulating noise with density matrices is memory intensive as it effectively
+doubles the number of qubits. Qibo provides an alternative way of simulating
+the effect of channels without using density matrices, which relies on state
+vectors and repeated circuit execution with sampling. Noise can be simulated
+by creating a normal (non-density matrix) circuit and repeating its execution
+as follows:
+
+.. code-block:: python
+
+    import numpy as np
+    from qibo import models, gates
+
+    # Define circuit
+    c = models.Circuit(5)
+    c.add((gates.RX(i, theta=t) for i, t in enumerate(np.random.random())))
```

Comment:
I think you have to set the random size explicitly otherwise the loop will fail.
```suggestion
    c.add((gates.RX(i, theta=t) for i, t in enumerate(np.random.random(1))))
```

Reply:
That's correct, thanks for noting. I fixed this together with the typo noted in #278 since it was the same file. It should be okay once we merge #278 here.